src/Features/CSharp/Portable/ConvertLinq/CSharpConvertLinqQueryToForEachProvider.cs

```diff
@@ -0,0 +1,792 @@
+                if (parentStatement != null)
+                {
+                    StatementSyntax internalNodeMethod(ExpressionSyntax expression)
+                        => SyntaxFactory.YieldStatement(SyntaxKind.YieldReturnStatement, expression);
+
+                    if (TryConvertQueryExpression(source, internalNodeMethod, out var convertedFromQueryExpression))
+                    {
+                        if (TryGetTypeSyntax(source, out var typeSyntax))
+                        {
+                            // before statement ... from a in select b ...
+                            // after
+                            // IEnumerable<T> localFunction()
+                            // {
+                            //   foreach(var a in b)
+                            //   {
+                            //       yield return a;
+                            //   }
+                            // }
+                            //  statement ... localFunction();
+                            string localFunctionName =_semanticFacts.GenerateNameForExpression(
+                                _semanticModel, 
+                                source, 
+                                capitalize: false, 
+                                _cancellationToken);
+                            localFunctionName = GetFreeSymbolName(localFunctionName, source.GetLocation().SourceSpan.Start);
+                            var localFunctionDeclaration = SyntaxFactory.LocalFunctionStatement(
+                                modifiers: default,
+                                returnType: typeSyntax.WithAdditionalAnnotations(Simplifier.SpecialTypeAnnotation),
+                                identifier: SyntaxFactory.Identifier(localFunctionName),
+                                typeParameterList: null,
+                                parameterList: SyntaxFactory.ParameterList(),
+                                constraintClauses: default,
+                                body: SyntaxFactory.Block(
+                                    SyntaxFactory.Token(
+                                        SyntaxFactory.TriviaList(),
+                                        SyntaxKind.OpenBraceToken,
+                                        SyntaxFactory.TriviaList(SyntaxFactory.EndOfLine(Environment.NewLine))),
+                                    SyntaxFactory.SingletonList(convertedFromQueryExpression),
+                                    SyntaxFactory.Token(SyntaxKind.CloseBraceToken)),
+                                expressionBody: null);
+
+                            var localFunctionInvocation = SyntaxFactory.InvocationExpression(SyntaxFactory.IdentifierName(localFunctionName)).WithAdditionalAnnotations(Simplifier.Annotation);
+                            SyntaxNode newParentExpressionStatement = parentStatement.ReplaceNode(source, localFunctionInvocation.WithAdditionalAnnotations(Simplifier.Annotation));
```

Comment:
>parentStatement [](start = 70, length = 15)

Could the ```parentStatement``` be the body of a lambda for example? Would the refactoring do the right that case?

Reply:
Yes, it could be a lambda. No issues should happen with this.

If we convert 
a => return query;
to yield return, there can be issues with yield return within lambda. Will catch this and introduce a local function within lambda (or anonymous method)

---
In reply to: [177166211](https://github.com/dotnet/roslyn/pull/25362#discussion_r177166211) [](ancestors = 177166211)